src/lib.rs

```diff
@@ -0,0 +1,698 @@
+    }
+}
+
+impl MemtestReport {
+    fn new(test_kind: MemtestKind, outcome: Result<MemtestOutcome, MemtestError>) -> MemtestReport {
+        MemtestReport { test_kind, outcome }
+    }
+}
+
+impl TimeoutChecker {
+    fn new(deadline: Instant) -> TimeoutChecker {
+        TimeoutChecker {
+            deadline,
+            state: None,
+        }
+    }
+
+    /// Initialize TimeoutCheckerState
+    /// This function should be called in the beginning of a memtest.
+    fn init(&mut self, expected_iter: u64) {
+        const FIRST_CHECKPOINT: u64 = 8;
+
+        assert!(
+            self.state.is_none(),
+            "init() should only be called once per test"
+        );
+
+        // The first checkpoint is set to 8 to have a more accurate sample of duration per
+        // iteration for determining new checkpoint
+        self.state = Some(TimeoutCheckerState {
+            test_start_time: Instant::now(),
+            expected_iter,
+            completed_iter: 0,
+            checkpoint: FIRST_CHECKPOINT,
+        });
+    }
+
+    /// Check if the current iteration is a checkpoint. If so, check if timeout occurred
+    ///
+    /// This function should be called in every iteration of a memtest.
+    ///
+    /// To reduce overhead, the function only checks for timeout at specific checkpoints, and
+    /// early returns otherwise.
+    ///
+    // It is important to ensure that the "early return" hot path is inlined. This results in a
+    // 100% improvement in performance.
+    #[inline(always)]
+    fn check(&mut self) -> Result<(), MemtestError> {
+        let state = self
+            .state
+            .as_mut()
+            .expect("init() should be called before check()");
+
+        if state.completed_iter < state.checkpoint {
+            state.completed_iter += 1;
+            return Ok(());
+        }
+
+        state.on_checkpoint(self.deadline)
+    }
+}
+
+impl TimeoutCheckerState {
+    fn on_checkpoint(&mut self, deadline: Instant) -> Result<(), MemtestError> {
+        self.trace_progress();
+
+        let current_time = Instant::now();
+        if current_time >= deadline {
+            return Err(MemtestError::Timeout);
+        }
+
+        self.set_next_checkpoint(deadline, current_time);
+
+        self.completed_iter += 1;
+        Ok(())
+    }
+
+    // Note: Because `trace_progress()` is only called in `on_checkpoints()`, not every percent
+    // of the test progress is traced. If memtests are running way ahead of the given deadline, the
+    // progress may only be traced once or twice. Although this makes the logs less comprehensive,
+    // it avoids signficiant perforamnce overhead.
+    fn trace_progress(&mut self) {
+        if tracing::enabled!(tracing::Level::TRACE) {
+            let hundredth_expected_iter = (self.expected_iter as f32 / 100.0).ceil() as u64;
+            if self.completed_iter % hundredth_expected_iter == 0 {
```

Comment:
I don't think this algorithm is going to work the way you want. The way you have it written, it will only log if you get lucky and accidently end up calling a checkpoint on an iteration that is a perfect percentage point. To demonstrate, let's assume `expected_iter == 1234`.

```Rust
let hundredth_expected_iter = (1234.00 / 100.0 = 12.34).ceil() = 13.00 as u64 = 13
if completed_iter % 13 == 0 {
    // Will only be run if this checkpoint happens to be perfectly divisable by 13, which may never happen.
}
```

I think you're left with roughly 3 choices:

1.  Print your progress unconditionally every single checkpoint. This is the only one that doesn't require additional state.
2.  Print your progress when it's been a certain amount of time since the last time you printed (IE print every 5 seconds). This requires you to store the last time you printed.
3.  Print your progress when it's been a certain number of iterations since the last time you printed (IE `hundredth_expected_iter`). This requires you to store the last iteration when you printed.

Reply:
Ah ok that makes sense. I think I'll just go with the 1.